tutorials/rust/harfbuzz/src/main.rs

```diff
@@ -0,0 +1,17 @@
+fn main() {
+    use harfbuzz::{sys, Buffer, Direction, UnicodeFuncsBuilder};
+    use icu_harfbuzz::AllUnicodeFuncs;
+    let mut b = Buffer::with("مساء الخير");
+    let mut builder = UnicodeFuncsBuilder::new_with_empty_parent().unwrap();
+    //  Note: AllUnicodeFuncs is zero-sized, so these boxes don't allocate memory.
+    builder.set_general_category_func(AllUnicodeFuncs::boxed());
+    builder.set_combining_class_func(AllUnicodeFuncs::boxed());
+    builder.set_mirroring_func(AllUnicodeFuncs::boxed());
+    builder.set_script_func(AllUnicodeFuncs::boxed());
+    builder.set_compose_func(AllUnicodeFuncs::boxed());
+    builder.set_decompose_func(AllUnicodeFuncs::boxed());
+    b.set_unicode_funcs(&builder.build());
+    b.guess_segment_properties();
```

Comment:
@sffc can we still make changes to the harfbuzz API? It would be nice if the builder setters were chainable:
```rust
    let funcs = UnicodeFuncsBuilder::new_with_empty_parent()
       .unwrap()
       .set_general_category_func(AllUnicodeFuncs::boxed())
       .set_combining_class_func(AllUnicodeFuncs::boxed())
       .set_mirroring_func(AllUnicodeFuncs::boxed())
       .set_script_func(AllUnicodeFuncs::boxed())
       .set_compose_func(AllUnicodeFuncs::boxed())
       .set_decompose_func(AllUnicodeFuncs::boxed())
       .build();
    let mut b = Buffer::with("مساء الخير");
    b.set_unicode_funcs(&funcs)
```

Reply:
This API was designed to closely mirror the C FFI. The "builder" is actually a real Harfbuzz C object. Every one of these setters calls FFI: it's not just a figment of Rust. The C FFI takes a mutable pointer and sets stuff on the mutable pointer, so the Rust API should by default do the same. The reason I called it a "builder", despite that term not appearing in the Harfbuzz docs, is because the C FFI has a function `hb_unicode_funcs_make_immutable` (which seems to be similar to the "freeze" concept in ICU4C), and this was the easiest way to track the state of whether the object was mutable or immutable.

https://harfbuzz.github.io/harfbuzz-hb-unicode.html

Now of course there is room to debate the API design in Rust; I took some liberty by calling the thing a "builder". However, I haven't seen evidence that my design has fatal flaws, and in fact this PR validates my approach. Changing between a mutating builder and a moving builder is a matter of debate that doesn't change what happens under the hood. And, the `harfbuzz` crate is external to ICU4X, living on its own lifecycle, and the maintainers of that crate already reviewed and had me change some aspects of the API before it landed. It is a crate designed for a narrow set of power users who know what they are doing. Most folks wanting to do text layout in Rust should hopefully be using a more end-to-end package instead of directly interfacing with harfbuzz.

For all these reasons, I would not personally be enthusiastic about a change of this nature,.